go/vt/vtgate/semantics/info_schema.go

```diff
@@ -1671,16 +1671,25 @@ type infoSchemaWithColumns struct {
 	infoSchemaData map[string][]vindexes.Column
 }
 
+// We cache this information. If the registered mysql version is changed, we flush the cache
```

Comment:
Can this happen for a `vtgate`? Afaik you can't change the value of `servenv.MySQLServerVersion()` at runtime? So it shouldn't be needed to guard for changing this? 

If it could be changed at runtime, I think this code is racy? Even though it assigns to the global variable to overwrite, it could still see partially initialized data then. See also the Go memory model under `Incorrect synchronization` at https://go.dev/ref/mem

Reply:
the mysql version cannot be changed at runtime. But the `getInfoSchema<version>` calls are slow and they are executed on every query. Catching this information is the right thing. We can remove the logic of changing this on runtime as it would not be needed.

Reply:
Let's build this information on `init`

Reply:
We do change this on some tests, as seen by the method `SetMySQLServerVersionForTest`.

WDYT - make sure it's not racy, if you think that's important for the tests. or is it enough to slap a big disclaimer saying that the code is racy and not meant to be changed in production?

Reply:
@systay If it's changed during tests, won't we have problems with potentially other tests running concurrently that might exercise this too and being racy that way?

Reply:
Alternatively, we can do a variation on what @harshit-gangal mentioned. So we always statically initialize both for 5.7 and 8.0 and at runtime we pick the cached version that we care about. That way we won't have the race issues either since both are initialized? 

```
SetMySQLServerVersionForTest
```

I think that one is problematic in general, but points at a design flaw we really have which is that this version is global like this but we do want to change it for tests. I think it could already be affecting / making test racy today anyway if we change globals like that on the fly while other stuff is running too.

Reply:
I followed your advice. This is at least not making anything worse than it was before